File: backend/config/cache.js
// =============================================================================
// config/cache.js  —  Redis-first, in-memory fallback
// =============================================================================
// • If REDIS_URL is set → uses ioredis (production)
// • Otherwise           → falls back to in-process Map (dev / single-server)
//
// All callers use the same interface:
//   cache.get(key)            → value | null
//   cache.set(key, val, ttlMs)
//   cache.del(key)
//   cache.delPrefix(prefix)
//   cache.flush()
//
// Cache key convention:
//   ct:<counseling_type_id>              counseling-type row
//   filters:<ctName>                     full filter blob for one type
//   facet:st:<ctId>:<stId>              courses for (ct, state)
//   facet:qu:<ctId>:<stId>:<crsId>      quotas for (ct, state, course)
//   colleges:<ctId>:<stId>:<crsId>:<qId>:<page>:<search>
//   eligibility:<ctId>:<stId>:<catId>:<crsId>:<rndId>:<rank>:<page>
//   import_version                       bumped after every import
// =============================================================================

let redisClient = null;

if (process.env.REDIS_URL) {
  try {
    const Redis = require('ioredis');
    redisClient = new Redis(process.env.REDIS_URL, {
      maxRetriesPerRequest: 3,
      enableReadyCheck: true,
      lazyConnect: false,
    });
    redisClient.on('connect',  () => console.log('✅ Redis connected'));
    redisClient.on('error',    (e) => console.error('❌ Redis error:', e.message));
  } catch (e) {
    console.warn('⚠️  ioredis not installed, falling back to memory cache:', e.message);
    redisClient = null;
  }
} else if (process.env.NODE_ENV === 'production') {
  // Important once you scale beyond a single Render instance: the in-memory
  // fallback below is per-process only. With multiple Render instances
  // serving traffic (which you'll need for 10L+ users), each instance keeps
  // its own separate cache — a filter result cached by instance A won't be
  // visible to instance B, so cache hit rate (and the speed-up it gives)
  // silently drops as you scale out. Add a Redis add-on (Render Redis,
  // Upstash, etc.) and set REDIS_URL to share one cache across all instances.
  console.warn('⚠️  REDIS_URL not set in production — using per-instance in-memory cache. ' +
    'This works fine on a single instance but will NOT be shared once you scale to ' +
    'multiple Render instances. Add Redis for consistent caching at scale.');
}

// ── In-memory fallback ────────────────────────────────────────────────────────
const store = new Map();

setInterval(() => {
  const now = Date.now();
  for (const [k, v] of store.entries()) {
    if (now > v.expiresAt) store.delete(k);
  }
}, 5 * 60_000);

// ── Unified interface ─────────────────────────────────────────────────────────
const cache = {
  async get(key) {
    if (redisClient) {
      try {
        const raw = await redisClient.get(key);
        return raw ? JSON.parse(raw) : null;
      } catch { return null; }
    }
    const entry = store.get(key);
    if (!entry) return null;
    if (Date.now() > entry.expiresAt) { store.delete(key); return null; }
    return entry.value;
  },

  async set(key, value, ttlMs = 600_000) {
    if (redisClient) {
      try {
        await redisClient.set(key, JSON.stringify(value), 'PX', ttlMs);
      } catch {}
      return;
    }
    store.set(key, { value, expiresAt: Date.now() + ttlMs });
  },

  async del(key) {
    if (redisClient) { try { await redisClient.del(key); } catch {} return; }
    store.delete(key);
  },

  async delPrefix(prefix) {
    if (redisClient) {
      try {
        const keys = await redisClient.keys(`${prefix}*`);
        if (keys.length) await redisClient.del(...keys);
      } catch {}
      return;
    }
    for (const k of store.keys()) {
      if (k.startsWith(prefix)) store.delete(k);
    }
  },

  async flush() {
    if (redisClient) { try { await redisClient.flushdb(); } catch {} return; }
    store.clear();
  },

  // Called by the import pipeline after every successful CSV run
  async bumpImportVersion() {
    const v = Date.now().toString();
    await this.set('import_version', v, 30 * 24 * 60 * 60_000); // 30 days
    await this.delPrefix('filters:');
    await this.delPrefix('facet:');
    await this.delPrefix('colleges:');
    await this.delPrefix('eligibility:');
    await this.delPrefix('ct:');
    console.log('✅ Cache invalidated after import, version:', v);
    return v;
  },

  async getImportVersion() {
    return (await this.get('import_version')) || '0';
  },
};

module.exports = cache;

File: python_backend/cache.py
# =============================================================
# python_backend/cache.py — Redis-first, in-memory fallback
# =============================================================
# Mirrors the Node backend's config/cache.js on purpose:
#   • If REDIS_URL is set   → uses redis-py (shared across all
#     uvicorn workers/processes — fixes the bug where each worker
#     used to have its OWN separate in-memory cache).
#   • Otherwise              → falls back to an in-process dict
#     (fine for local dev / a single worker).
#
# Interface (all sync, since main.py routes are sync `def`, not
# `async def` — matches the existing SQLAlchemy sync session usage):
#   cache.get(key)                  -> value | None
#   cache.set(key, value, ttl_seconds=3600)
#   cache.delete_prefix(prefix)
#   cache.flush()
#   cache.bump_import_version()     -> call this (or hit
#                                       POST /api/admin/cache-invalidate)
#                                       right after a CSV import finishes.
#   cache.get_import_version()
# =============================================================

import os
import time
import json
import threading
import logging

logger = logging.getLogger(__name__)

_redis_client = None
REDIS_URL = os.getenv("REDIS_URL")

if REDIS_URL:
    try:
        import redis as _redis_pkg
        _redis_client = _redis_pkg.Redis.from_url(REDIS_URL, decode_responses=True)
        _redis_client.ping()
        logger.info("✅ Redis connected (python backend)")
    except Exception as e:
        logger.warning("⚠️  Redis unavailable, falling back to memory cache: %s", e)
        _redis_client = None

# ── In-memory fallback ────────────────────────────────────────
_store: dict = {}
_lock = threading.Lock()


def get(key: str):
    if _redis_client:
        try:
            raw = _redis_client.get(key)
            return json.loads(raw) if raw else None
        except Exception:
            return None
    with _lock:
        entry = _store.get(key)
        if not entry:
            return None
        if time.time() > entry["expires_at"]:
            del _store[key]
            return None
        return entry["value"]


def set(key: str, value, ttl_seconds: int = 3600):
    if _redis_client:
        try:
            _redis_client.set(key, json.dumps(value), ex=ttl_seconds)
        except Exception as e:
            logger.warning("Redis SET failed for %s: %s", key, e)
        return
    with _lock:
        if len(_store) > 5000:
            _store.clear()
        _store[key] = {"value": value, "expires_at": time.time() + ttl_seconds}


def delete_prefix(prefix: str):
    if _redis_client:
        try:
            keys = _redis_client.keys(f"{prefix}*")
            if keys:
                _redis_client.delete(*keys)
        except Exception as e:
            logger.warning("Redis delete_prefix failed for %s: %s", prefix, e)
        return
    with _lock:
        for k in list(_store.keys()):
            if k.startswith(prefix):
                del _store[k]


def flush():
    if _redis_client:
        try:
            _redis_client.flushdb()
        except Exception as e:
            logger.warning("Redis flush failed: %s", e)
        return
    with _lock:
        _store.clear()


def bump_import_version() -> str:
    """
    Call this right after uploader_turbo_v3.py finishes a CSV import
    (or POST /api/admin/cache-invalidate with the shared secret), so
    stale predictions/upgrade-checks are never served after data
    refreshes. Mirrors cache.js's bumpImportVersion().
    """
    v = str(int(time.time() * 1000))
    set("import_version", v, ttl_seconds=30 * 24 * 60 * 60)  # 30 days
    delete_prefix("optimize:")
    delete_prefix("upgrade:")
    delete_prefix("filters:")
    logger.info("✅ Python backend cache invalidated, version: %s", v)
    return v


def get_import_version() -> str:
    return get("import_version") or "0"
